### PR TITLE
Avoid capturing logger pointer in callback lambda

### PR DESCRIPTION
**Pull request by bianpengyuan**
Also add mesh_uid into metric and log labels.

### CODE REVIEW

File: extensions/stackdriver/log/exporter.cc [hunk 1]
```diff
@@ -42,19 +42,23 @@ namespace Stackdriver {
...
-  success_callback_ = [this](google::protobuf::Empty&&) {
-    export_call_.increment(1, "logging", true);
+  Metric export_call(MetricType::Counter, "stackdriver_filter",
+                     {MetricTag{"type", MetricTag::TagType::String},
+                      MetricTag{"success", MetricTag::TagType::Bool}});
+  auto success_counter = export_call.resolve("logging", true);
+  auto failure_counter = export_call.resolve("logging", false);
+  success_callback_ = [success_counter](google::protobuf::Empty&&) {
+    // TODO(bianpengyuan): replace this with envoy's generic gRPC counter.
+    incrementMetric(success_counter, 1);
```
Comment (mandarjog) [inline extensions/stackdriver/log/exporter.cc]: @jplevyak this was our solution to simplify the callback by not capturing references.
This lambda just captures the returned `int` and will use the C API to call into the host.
What is the behaviour if `incrementMetric` is called with an invalid id ?

File: extensions/stackdriver/log/logger_test.cc [hunk 2]
```diff
@@ -98,6 +99,7 @@ google::logging::v2::WriteLogEntriesRequest expectedRequest(
   (*top_label_map)["destination_name"] = node_info.name();
   (*top_label_map)["destination_workload"] = node_info.workload_name();
   (*top_label_map)["destination_namespace"] = node_info.namespace_();
+  (*top_label_map)["mesh_uid"] = node_info.mesh_id();
```
Comment (kyessenov) [inline extensions/stackdriver/log/logger_test.cc]: This MESH_ID change should be a separate (small) PR.

File: extensions/stackdriver/log/exporter.cc [hunk 3]
```diff
@@ -42,19 +42,23 @@ namespace Stackdriver {
...
     logDebug("successfully sent Stackdriver logging request");
   };
 
-  failure_callback_ = [this](GrpcStatus status, StringView message) {
+  failure_callback_ = [failure_counter](GrpcStatus status, StringView message) {
     // TODO(bianpengyuan): add retry.
-    export_call_.increment(1, "logging", false);
```
Comment (kyessenov) [inline extensions/stackdriver/log/exporter.cc]: (cc @jplevyak) I think we need a regression test for this, but I don't see a simple way to simulate this in envoy test framework. Let's file an issue to create a test case for reproducing the failure to confirm this fix works.

File: extensions/stackdriver/log/exporter.cc (hunk 1 above)
Comment (jplevyak) [reply extensions/stackdriver/log/exporter.cc]: The wasm container stores a map from int -> metric and does not free any metrics during the lifetime of the VM, so a valid int for a metric will remain so as long as the VM lives.

File: extensions/stackdriver/log/exporter.cc (hunk 3 above)
Comment (jplevyak) [reply extensions/stackdriver/log/exporter.cc]: What are you trying to simulate?  This failure was caused by two onConfigure calls coming in.  That can be simulated with the existing test framworks although it is only meaningful in the context of a particular plugin which many or may not handle it correctly.

Comment (bianpengyuan) [reply extensions/stackdriver/log/exporter.cc]: Done. https://github.com/istio/istio/issues/18396

File: extensions/stackdriver/log/logger_test.cc (hunk 2 above)
Comment (bianpengyuan) [reply extensions/stackdriver/log/logger_test.cc]: Done. https://github.com/istio/proxy/pull/2494

**Comment by mandarjog** [comment]
@kyessenov PTAL.

@bianpengyuan asan build is failing.

**Comment by bianpengyuan** [comment]
@mandarjog Hmm, it does not look related to my change. Let me try to rerun.

**Pull request merged**